### ISSUE TITLE
inet_dns:decode/1 srv and txt decode regression in 24.1.2

### ISSUE THREAD

**Issue by fhunleth** [erlang/otp#5327]
**Describe the bug**

Sometimes SRV and TXT data is not decoded with OTP 24.1.2. It works right with OTP 24.1.1. See the example since there's something about the data that's causing the appropriate decode function to not be called.

**To Reproduce**

Here's a good run with OTP 24.1.1:

```erl
1> inet_dns:decode(<<0, 0, 132, 0, 0, 0, 0, 1, 0, 0, 0, 6, 4, 95, 101, 108, 103, 4, 95, 116, 99, 112, 5, 108, 111, 99, 97, 108, 0, 0, 12, 0, 1, 0, 0, 17, 148, 0, 24, 21, 69, 108, 103, 97, 116, 111, 32, 75, 101, 121, 32, 76, 105, 103, 104, 116, 32, 57, 57, 51, 66, 192, 12, 21, 101, 108, 103, 97, 116, 111, 45, 107, 101, 121, 45, 108, 105, 103, 104, 116, 45, 57, 57, 51, 98, 192, 22, 0, 1, 128, 1, 0, 0, 0, 120, 0, 4, 192, 168, 3, 39, 192, 63, 0, 28, 128, 1, 0, 0, 0, 120, 0, 16, 254, 128, 0, 0, 0, 0, 0, 0, 62, 106, 157, 255, 254, 20, 213, 105, 192, 39, 0, 33, 128, 1, 0, 0, 0, 120, 0, 8, 0, 0, 0, 0, 35, 163, 192, 63, 192, 39, 0, 16, 128, 1, 0, 0, 17, 148, 0, 74, 9, 109, 102, 61, 69, 108, 103, 97, 116, 111, 5, 100, 116, 61, 53, 51, 20, 105, 100, 61, 51, 67, 58, 54, 65, 58, 57, 68, 58, 49, 52, 58, 68, 53, 58, 54, 57, 29, 109, 100, 61, 69, 108, 103, 97, 116, 111, 32, 75, 101, 121, 32, 76, 105, 103, 104, 116, 32, 50, 48, 71, 65, 75, 57, 57, 48, 49, 6, 112, 118, 61, 49, 46, 48, 192, 63, 0, 47, 128, 1, 0, 0, 0, 120, 0, 8, 192, 63, 0, 4, 64, 0, 0, 8, 192, 39, 0, 47, 128, 1, 0, 0, 0, 120, 0, 9, 192, 39, 0, 5, 0, 0, 128, 0, 64>>).
{ok,{dns_rec,{dns_header,0,true,query,true,false,false,
                         false,false,0},
             [],
             [{dns_rr,"_elg._tcp.local",ptr,in,0,4500,
                      "Elgato Key Light 993B._elg._tcp.local",undefined,[],false}],
             [],
             [{dns_rr,"elgato-key-light-993b.local",a,32769,0,120,
                      <<192,168,3,39>>,
                      undefined,[],false},
              {dns_rr,"elgato-key-light-993b.local",aaaa,32769,0,120,
                      <<254,128,0,0,0,0,0,0,62,106,157,255,...>>,
                      undefined,[],false},
              {dns_rr,"Elgato Key Light 993B._elg._tcp.local",srv,32769,0,
                      120,
                      {0,0,9123,"elgato-key-light-993b.local"},
                      undefined,[],false},
              {dns_rr,"Elgato Key Light 993B._elg._tcp.local",txt,32769,0,
                      4500,
                      ["mf=Elgato","dt=53","id=3C:6A:9D:14:D5:69",
                       "md=Elgato Key Light 20GAK9901","pv=1.0"],
                      undefined,[],false},
              {dns_rr,"elgato-key-light-993b.local",47,32769,0,120,
                      <<192,63,0,4,64,0,0,8>>,
                      undefined,[],false},
              {dns_rr,"Elgato Key Light 993B._elg._tcp.local",47,32769,0,
                      120,
                      <<192,39,0,5,0,0,128,0,...>>,
                      undefined,[],false}]}}
```

The important part is the `srv` and `txt` records. The data is nicely decoded into a tuple and list. 

On OTP 24.1.2, here's the output:

```erlang
2> inet_dns:decode(<<0, 0, 132, 0, 0, 0, 0, 1, 0, 0, 0, 6, 4, 95, 101, 108, 103, 4, 95, 116, 99, 112, 5, 108, 111, 99, 97, 108, 0, 0, 12, 0, 1, 0, 0, 17, 148, 0, 24, 21, 69, 108, 103, 97, 116, 111, 32, 75, 101, 121, 32, 76, 105, 103, 104, 116, 32, 57, 57, 51, 66, 192, 12, 21, 101, 108, 103, 97, 116, 111, 45, 107, 101, 121, 45, 108, 105, 103, 104, 116, 45, 57, 57, 51, 98, 192, 22, 0, 1, 128, 1, 0, 0, 0, 120, 0, 4, 192, 168, 3, 39, 192, 63, 0, 28, 128, 1, 0, 0, 0, 120, 0, 16, 254, 128, 0, 0, 0, 0, 0, 0, 62, 106, 157, 255, 254, 20, 213, 105, 192, 39, 0, 33, 128, 1, 0, 0, 0, 120, 0, 8, 0, 0, 0, 0, 35, 163, 192, 63, 192, 39, 0, 16, 128, 1, 0, 0, 17, 148, 0, 74, 9, 109, 102, 61, 69, 108, 103, 97, 116, 111, 5, 100, 116, 61, 53, 51, 20, 105, 100, 61, 51, 67, 58, 54, 65, 58, 57, 68, 58, 49, 52, 58, 68, 53, 58, 54, 57, 29, 109, 100, 61, 69, 108, 103, 97, 116, 111, 32, 75, 101, 121, 32, 76, 105, 103, 104, 116, 32, 50, 48, 71, 65, 75, 57, 57, 48, 49, 6, 112, 118, 61, 49, 46, 48, 192, 63, 0, 47, 128, 1, 0, 0, 0, 120, 0, 8, 192, 63, 0, 4, 64, 0, 0, 8, 192, 39, 0, 47, 128, 1, 0, 0, 0, 120, 0, 9, 192, 39, 0, 5, 0, 0, 128, 0, 64>>).
{ok,{dns_rec,{dns_header,0,true,query,true,false,false,
                         false,false,0},
             [],
             [{dns_rr,"_elg._tcp.local",ptr,in,0,4500,
                      "Elgato Key Light 993B._elg._tcp.local",undefined,[],false}],
             [],
             [{dns_rr,"elgato-key-light-993b.local",a,32769,0,120,
                      <<192,168,3,39>>,
                      undefined,[],false},
              {dns_rr,"elgato-key-light-993b.local",aaaa,32769,0,120,
                      <<254,128,0,0,0,0,0,0,62,106,157,255,...>>,
                      undefined,[],false},
              {dns_rr,"Elgato Key Light 993B._elg._tcp.local",srv,32769,0,
                      120,
                      <<0,0,0,0,35,163,192,63>>,
                      undefined,[],false},
              {dns_rr,"Elgato Key Light 993B._elg._tcp.local",txt,32769,0,
                      4500,<<"\tmf=Elgato"...>>,undefined,[],false},
              {dns_rr,"elgato-key-light-993b.local",47,32769,0,120,
                      <<192,63,0,4,64,0,0,8>>,
                      undefined,[],false},
              {dns_rr,"Elgato Key Light 993B._elg._tcp.local",47,32769,0,
                      120,
                      <<192,39,0,5,0,0,128,0,...>>,
                      undefined,[],false}]}}
```

The data in the `srv` and `txt` records is undecoded.

**Expected behavior**

The `srv` and `txt` data is decoded.

**Affected versions**

OTP 24.1.1 and 24.1 both work as expected. It looks like the update to `inet_dns.erl` for OTP 24.1.2 has something to do with the change, but I'm not seeing it yet,

**Comment by RaimoNiskanen** [comment]
Well, that's awkward...

The mDNS [RFC6762](https://datatracker.ietf.org/doc/html/rfc6762) invalidates the assumption that it is suitable to represent the Class as either a known atom or a raw integer, but instead extends the known classes with a flag bit.

Since for example A and AAAA records are only defined for class IN we need awareness of this flag bit, or rather of that it is class IN also in presence of the flag bit; so we will be able to use the non-raw address format `{A,B,C,D}` for IN, A and the corresponding for AAAA.

Today a numeric class, or a numeric type indicates that the record does not have a known format, even if the numerical values are for a known class and/or type.  With this approach it is possible to use binary encoding of the RR content also for known record classes and types.

The simplest and non-pretty solution I can think of right now is to make this a bit non-symmetric, i.e to allow and recognize numeric class also for known classes (with or without the high flag bit in mDNS), and a numeric type when encoding would be the only thing that selects binary encoding.
...and when decoding keep the class as numeric when it has a high flag bit, and use an atom type with decoded content if the class is of a known numeric value.

Would this be a good enough solution?

**Comment by fhunleth** [comment]
Your solution is good enough for my use. Thank you!

**Comment by RaimoNiskanen** [comment]
A "prettier" alternative would be to introduce values such as `{Class, unicast_response}` for `#dns_query.class` and `{Class, cache_flush}` for `#dns_rr.class`.  Then code that receives such queries would need to be modified to handle them.
Perhaps that is not much more backwards incompatible, since my "good enough" solution above would require code to be changed to accept a numeric class combined with symbolic type and decoded data.

**Comment by RaimoNiskanen** [comment]
@fhunleth: Would the latter solution also be acceptable?

**Comment by fhunleth** [comment]
Overall, I'm good with any of the solutions, since I feel like 1.) it's not hard to adapt to either of the solutions, 2.) `inet_dns:decode/1` is an undocumented API call, so I shouldn't expect it to be backwards compatible, and 3.) I'm probably going to have to vendor the `inet_dns:decode/1` implementation in the near-term until whatever gets decided upon is released.

Having said that, I assume you've thought about adding a `unicast_response` field to the `dns_query` record? And then have the `class` be `in` regardless of the unicast response flag.

**Comment by RaimoNiskanen** [comment]
> Having said that, I assume you've thought about adding a unicast_response field to the `#dns_query` record? And then have the class be in regardless of the unicast response flag.

I did think about that, it feels like the "best" solution, but a field would be needed in the `#dns_rr` record too (that would be carried in to the RR cache ETS table).

The `#dns_rr` record is "more public" than the `inet_dns` module.  There are access functions in `inet_dns` documented in `inet_res`, and the record itself is stated as opaque.  But everybody know how to include the header file and how to use the record directly.
Therefore I am a little afraid of how much code "out there" that would need recompiling to work at all with the new record format...

But then again, still formally undocumented...

**Comment by fhunleth** [comment]
I'm happy with your solution in #5338. I'll try it out with I get a window of time later today or tomorrow morning just to be sure. Thank you for addressing that so quickly.

Thanks for the comments about adding a field.

**Comment by RaimoNiskanen** [comment]
A "creative" solution, that I thought of last night, and that might be better; would be to add a field as you suggested to the `#dns_query` record, and to re-use the `#dns_rr.func` field that I never have seen being used as it is commented "Optional function calculating the data field", and it has luckily the default value `false` which fits very well.

On the master branch I have already changed meaning of the `#dns_rr.bm` field ("Bitmap storing domain character case information") into "Case normalized domain" to optimize ETS lookups, and so far seen no problems.

**Comment by RaimoNiskanen** [comment]
I pushed the latest solution above to PR #5338 instead, and think that it is a better one.

**Comment by RaimoNiskanen** [comment]
Fix in PR #5338 released in OTP-24.1.5.

Thank you for reporting this!
